### PR TITLE
provenance: Add web end point

### PR DESCRIPTION
**Pull request by ghost**
https://github.com/openactive/conformance-services/issues/51

### CODE REVIEW

**Comment by odscjames** [comment]
Here's a sample.

It contains the raw item and the parent raw item. It even contains raw data validation results, if they exist!

Technically both raw items should be from the same publisher I think, but this will cover any eventuality (like me bodging the DB to produce this one) and provides all the info, which is the main point.


```

{
  "id": "17dc76c2b1fd0aa61a676c91e356349b087c12b0",
  "deleted": false,
  "data": {
    "@id": "https://visionrcl.leisurecloud.net/OpenActive/api/scheduled-sessions/20449958",
    "@type": "ScheduledSession",
    "endDate": "2020-07-29T10:00:00Z",
    "@context": [
      "https://openactive.io/",
      "https://openactive.io/ns-beta"
    ],
    "duration": "PT1H",
    "startDate": "2020-07-29T09:00:00Z",
    "identifier": 20449958,
    "superEvent": "https://visionrcl.leisurecloud.net/OpenActive/api/session-series/FC8GYMB10000720",
    "maximumAttendeeCapacity": 42,
    "remainingAttendeeCapacity": 42,
    "beta:sportsActivityLocation": [
      {
        "name": "1st Floor",
        "type": "SportsActivityLocation"
      }
    ]
  },
  "updated_at": "2020-07-24T13:27:38.259Z",
  "kind": "ScheduledSession",
  "raw_data": {
    "rpde_id": "00000000000020449958",
    "data_id": "https://visionrcl.leisurecloud.net/OpenActive/api/scheduled-sessions/20449958",
    "kind": "ScheduledSession",
    "data": {
      "id": "https://visionrcl.leisurecloud.net/OpenActive/api/scheduled-sessions/20449958",
      "type": "ScheduledSession",
      "endDate": "2020-07-29T10:00:00Z",
      "@context": [
        "https://openactive.io/",
        "https://openactive.io/ns-beta"
      ],
      "duration": "PT1H",
      "startDate": "2020-07-29T09:00:00Z",
      "identifier": 20449958,
      "superEvent": "https://visionrcl.leisurecloud.net/OpenActive/api/session-series/FC8GYMB10000720",
      "maximumAttendeeCapacity": 42,
      "remainingAttendeeCapacity": 42,
      "beta:sportsActivityLocation": [
        {
          "name": "1st Floor",
          "type": "SportsActivityLocation"
        }
      ]
    },
    "validation": {
      "done": false
    },
    "feed": {
      "name": "ScheduledSession",
      "url": "https://opendata.leisurecloud.live/api/feeds/VisionRedbridge-live-scheduled-sessions",
      "publisher": {
        "name": "Vision Redbridge Culture & Leisure Ltd Sessions and Facilities",
        "url": "https://visionrcl.leisurecloud.net/OpenActive/"
      }
    }
  },
  "parent_raw_data": {
    "rpde_id": "00000000000070959710",
    "data_id": "https://api.activenewham.org.uk/OpenActive/api/scheduled-sessions/70959710",
    "kind": "ScheduledSession",
    "data": {
      "id": "https://api.activenewham.org.uk/OpenActive/api/scheduled-sessions/70959710",
      "type": "ScheduledSession",
      "endDate": "2020-08-06T19:00:00Z",
      "@context": [
        "https://openactive.io/",
        "https://openactive.io/ns-beta"
      ],
      "duration": "PT1H",
      "startDate": "2020-08-06T18:00:00Z",
      "identifier": 70959710,
      "superEvent": "https://api.activenewham.org.uk/OpenActive/api/session-series/N4CLKEY19001217",
      "maximumAttendeeCapacity": 40,
      "remainingAttendeeCapacity": 40,
      "beta:sportsActivityLocation": [
        {
          "name": "Studio 1",
          "type": "SportsActivityLocation"
        }
      ]
    },
    "validation": {
      "done": true,
      "results": [
        {
          "path": "$[\"beta:sportsActivityLocation\"]",
          "rule": "FieldsNotInModelRule",
          "type": "experimental_fields_not_checked",
          "value": [
            {
              "name": "Studio 1",
              "type": "SportsActivityLocation"
            }
          ],
          "message": "A definition for this extension property was found, but it has not been included in the correct object type. Please check the spelling of this property and ensure that you are using it within the correct object `\"type\"`.\n\nThe types allowed for this property are:\n\n<ul><li>https://schema.org/Event</li><li>https://openactive.io/Slot</li><li>https://openactive.io/FacilityUse</li></ul>\n\nFor more information about extension properties, see the [extension properties guide](https://openactive.io/modelling-opportunity-data/EditorsDraft/#defining-and-using-custom-namespaces).",
          "category": "conformance",
          "severity": "failure"
        }
      ]
    },
    "feed": {
      "name": "ScheduledSession",
      "url": "https://opendata.leisurecloud.live/api/feeds/ActiveNewham-live-live-scheduled-sessions",
      "publisher": {
        "name": "activeNewham Sessions and Facilities",
        "url": "https://api.activenewham.org.uk/OpenActive/"
      }
    }
  }
}
```

**Comment by odscjames** [comment]
[Add provenanceInformation block](https://trello.com/c/c35PCWAy/125-add-provenanceinformation-block)

File: src/lib/web-server.js [hunk 1]
```diff
@@ -49,6 +50,19 @@ web_server_app.get('/normalised_data/all', async (req, res) => {
     res.json(out);
 });
 
+
+web_server_app.get('/normalised_data/item_by_id/:data_id', async (req, res) => {
```
Comment (michaelwood) [inline src/lib/web-server.js:54]: Minor thing (not blocking) but can we drop "_by_id" that feels a little cumbersome. "/normalised_data/item/:data_id" looks better.

File: src/lib/web-normalised-data-info.js [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        *
+      FROM publisher
+      WHERE id=$1
+    `;
+
+    const res_raw_data = await client.query(raw_data_sql_query, [raw_data_id]);
+    const raw_data = res_raw_data.rows[0];
+
+    let out = {
+        'rpde_id': raw_data.rpde_id,
```
Comment (michaelwood) [inline src/lib/web-normalised-data-info.js:79]: can we use camelCase for keys in our JSON as that makes it consistent with other bits of conformance-services and OpenActive generally

**Comment by michaelwood** [comment]
I've added a commit to address those issues.

File: src/lib/web-normalised-data-info.js [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        out = {
+            'id': normalised_data.data_id,
+            'deleted': normalised_data.data_deleted,
+            'data': normalised_data.data,
+            'updated_at': normalised_data.updated_at,
+            'kind': normalised_data.data_kind
+        };
+
+        out['raw_data'] = await this.get_raw_data_info(normalised_data.raw_data_id, client);
+
```
Comment (michaelwood) [inline src/lib/web-normalised-data-info.js:30]: normally in JS you'd just do `out.raw_data = ...etc`  you generally only use the assoc array stuff if you have a variable as a name e.g. `let key = "moo"; out[key] = ...etc`

**Review by rhiaro — APPROVED** [review]

**Pull request merged**